File: python/symjit/pyengine/assembler.py
class Assembler:
    def __init__(self):
        self.buf = bytearray()
        self.tail = None
        self.labels = {}
        self.jumps = []

    def append_byte(self, *b):
        self.buf.extend(b)

    def append_word(self, u):
        """appends u (uint32) as little-endian"""
        self.buf.append(u & 0xFF)
        self.buf.append((u >> 8) & 0xFF)
        self.buf.append((u >> 16) & 0xFF)
        self.buf.append((u >> 24) & 0xFF)

    def begin_prepend(self):
        assert self.tail is None
        self.tail = self.buf
        self.buf = bytearray()

    def end_prepend(self):
        assert self.tail is not None
        shift = len(self.buf)
        self.buf.extend(self.tail)
        self.tail = None
        # note: prologue cannot have jumps
        jumps = []
        for label, k in self.jumps:
            jumps.append((label, k + shift))
        self.jumps = jumps

    def set_label(self, label):
        self.labels[label] = len(self.buf)

    def jump(self, label, code=0):
        self.jumps.append((label, len(self.buf)))
        self.append_word(code)

    def test(self, b):
        if self.buf == bytes(b):
            print(f"pass: {b}")
            self.buf.clear()
            return True
        else:
            return False

File: rust/assembler.rs
use std::collections::HashMap;

pub struct Assembler {
    buf: Vec<u8>,
    labels: HashMap<String, usize>,
    jumps: Vec<(String, usize, u32)>,
    delta: isize,
    shift: isize,
}

impl Assembler {
    pub fn new(delta: isize, shift: isize) -> Assembler {
        Assembler {
            buf: Vec::new(),
            labels: HashMap::new(),
            jumps: Vec::new(),
            delta,
            shift,
        }
    }

    pub fn bytes(&self) -> Vec<u8> {
        self.buf.clone()
    }

    pub fn append_byte(&mut self, b: u8) {
        self.buf.push(b)
    }

    pub fn append_bytes(&mut self, bs: &[u8]) {
        for b in bs {
            self.append_byte(*b);
        }
    }

    pub fn append_word(&mut self, mut u: u32) {
        // appends u (uint32) as little-endian
        for _ in 0..4 {
            self.append_byte((u & 0xff) as u8);
            u >>= 8;
        }
    }

    pub fn append_quad(&mut self, mut u: u64) {
        // appends u (uint32) as little-endian
        for _ in 0..8 {
            self.append_byte((u & 0xff) as u8);
            u >>= 8;
        }
    }

    pub fn ip(&self) -> usize {
        self.buf.len()
    }

    pub fn set_label(&mut self, label: &str) {
        self.labels.insert(label.to_string(), self.ip());
    }

    pub fn jump(&mut self, label: &str, code: u32) {
        self.jumps.push((label.to_string(), self.ip(), code));
        self.append_word(code);
    }

    pub fn apply_jumps(&mut self) {
        for (label, k, code) in self.jumps.iter() {
            let target = self.labels.get(label).expect("label not found");
            let offset = (*target as isize) - (*k as isize) + self.delta;
            let x = ((offset as u32) << self.shift) | *code;

            self.buf[*k] |= (x & 0xff) as u8;
            self.buf[*k + 1] |= ((x >> 8) & 0xff) as u8;
            self.buf[*k + 2] |= ((x >> 16) & 0xff) as u8;
            self.buf[*k + 3] |= ((x >> 24) & 0xff) as u8;
        }
    }
}
